addons/controllers/csi/csiconfig_controller.go

```diff
@@ -0,0 +1,196 @@
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *CSIConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("CSIConfig", req.NamespacedName)
+
+	csiConfig := &csiv1alpha1.CSIConfig{}
+	if err := r.Get(ctx, req.NamespacedName, csiConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("CSIConfig resource not found", "namespacedName", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Unable to fetch CSIConfig resource", "namespacedName", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	if cluster, err := r.getOwnerCluster(ctx, csiConfig); cluster == nil {
+		return ctrl.Result{}, err // no need to requeue if cluster is not found
+	}
+
+	return r.reconcileCSIConfig(ctx, csiConfig)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CSIConfigReconciler) SetupWithManager(_ context.Context, mgr ctrl.Manager,
+	options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1alpha1.CSIConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+func (r *CSIConfigReconciler) reconcileCSIConfig(ctx context.Context,
+	csiCfg *csiv1alpha1.CSIConfig) (result ctrl.Result, retErr error) {
+
+	logger := log.FromContext(ctx)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(csiCfg, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if retErr != nil {
+			// don't modify CSIConfig if there is an error
+			return
+		}
+
+		if err := patchHelper.Patch(ctx, csiCfg); err != nil {
+			logger.Error(err, "Failed to patch CSIConfig")
+			return
+		}
+	}()
+
+	if !csiCfg.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil // deleted
+	}
+
+	if result, err = r.reconcileCSIConfigNormal(ctx, csiCfg); err != nil {
+		logger.Error(err, "Error reconciling CSIConfig")
+		return result, err
+	}
+
+	return result, nil
+}
+
+func (r *CSIConfigReconciler) reconcileCSIConfigNormal(ctx context.Context,
+	csiCfg *csiv1alpha1.CSIConfig) (result ctrl.Result, retErr error) {
+
+	logger := log.FromContext(ctx)
+
+	// add owner reference to CSIConfig if not already added
+	cluster, err := r.getOwnerCluster(ctx, csiCfg)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	ownerRef := metav1.OwnerReference{
+		APIVersion: clusterapiv1beta1.GroupVersion.String(),
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+
+	if !clusterapiutil.HasOwnerRef(csiCfg.OwnerReferences, ownerRef) {
+		csiCfg.OwnerReferences = clusterapiutil.EnsureOwnerRef(csiCfg.OwnerReferences, ownerRef)
+	}
+
+	addonName := ""
+	switch csiCfg.Spec.VSphereCSI.Mode {
+	case CSINonParavirtualMode:
+		addonName = "vsphere-csi"
```

Comment:
`vsphere-csi` and `vsphere-pv-csi` can be stored as constants

Reply:
updated